### PR TITLE
[DO NOT MERGE] Upgrade terraform provider google beta to v5.0.0

### LINKED ISSUE #1187: Only-set-bigquery-dataset-selflink-if-available.patch
[This patch](https://github.com/pulumi/pulumi-gcp/blob/6a64ebe/patches/0003-Only-set-bigquery-dataset-selflink-if-available.patch) makes an action conditional on a property being present.

Presumably, this is not necessary when the provider is run against terraform, indicating a bridge bug. I looked but could not find an upstream issue.

### Tests
None

### Recommendation
If upstream removes the `self_link` field, this will introduce the field back into the resource, potentially erroring out. We should confirm that this patch is still necessary, and make sure that `resource_bigquery_dataset` is under test before the next major release.

### LINKED ISSUE #1189: Remove-security-policy-default-versioned_expr.patch
[This patch](https://github.com/pulumi/pulumi-gcp/blob/6a64ebe/patches/0005-Remove-security-policy-default-versioned_expr.patch) removes `Default: ""` from a field. Presumably this is because it is not compatible with the field `validator: ValidateFunc: validation.StringInSlice([]string{"SRC_IPS_V1"}, false).` We should confirm that this is still an issue and open a tracking issue in the bridge. This is almost certainly a bridge bug.

### Tests
None

### Recommendation
If the upstream provider assumes that a field is non-empty but we removed the default, we could see a panic. This should have a test before the next major release.

### LINKED ISSUE #1229: Upgrade terraform-provider-google-beta to v5.0.0
Release details: https://github.com/hashicorp/terraform-provider-google-beta/releases/tag/v5.0.0

### PR DESCRIPTION
**Pull request by guineveresaenger**
This pull request presents a major version upgrade for the Pulumi Google Provider. This work is tracked in #1183.

- Removes deprecated resources and hard aliases; see schema-diff summary below.
- Removes security policy patch as fixed via [the bridge](https://github.com/pulumi/pulumi-gcp/issues/1029#issuecomment-1472632119) and adds integration test to verify it works as expected
- Addresses [bigquery.dataset patch](https://github.com/pulumi/pulumi-gcp/issues/1187) as still needed and adds test to verify correct behavior
- Adjusts patch set for upstream changes
- Tests remain enabled for major version upgrade 🎉 
- Fixes `bucket-go` tests

Ready for review; waiting on [bridge update](https://github.com/pulumi/pulumi-gcp/pull/1256) and [upgrade tests](https://github.com/pulumi/pulumi-gcp/pull/1237) to merge and rebase.

Recommendation: When merged, cut alpha release and pivot to #1248.

Fixes #1229
Fixes #1189
Fixes #1187

### CODE REVIEW

**Comment by t0yv0** [comment]
Couple of preliminary notes... 

- we'll need an explanation for removed resources and functions; were those removed upstream? were these the hard aliases? there's a few so we might need to divide-and-conquer chasing these down - I see you mention "Removes deprecated resources and hard aliases; see schema-diff summary below." but we might need the detail for the migration guide? Perhaps if schema-diff can distinguish "removed hard alias" or some such we could quickly fix it up to give a more informative summary to save work here?

- following how we did the AWS release I think we should be able to cut alphas from a release branch, so we can verify them in templates/examples before merging

- do any of the changes like removing patches easily separable into their own PRs so we can check them onto main branch or that's not worth it?

File: patches/0004-rebase-bigquery_dataset.patch [hunk 1]
```diff
@@ -1,20 +1,27 @@
...
-index 96cf5c7b6..7086541bd 100644
+index d09ee5b55..e277b5cc1 100644
 --- a/google-beta/services/bigquery/resource_bigquery_dataset.go
 +++ b/google-beta/services/bigquery/resource_bigquery_dataset.go
-@@ -654,8 +654,13 @@ func resourceBigQueryDatasetRead(d *schema.ResourceData, meta interface{}) error
+@@ -676,14 +676,20 @@ func resourceBigQueryDatasetRead(d *schema.ResourceData, meta interface{}) error
  	if err := d.Set("storage_billing_model", flattenBigQueryDatasetStorageBillingModel(res["storageBillingModel"], d, config)); err != nil {
  		return fmt.Errorf("Error reading Dataset: %s", err)
  	}
++
```
Comment (t0yv0) [inline patches/0004-rebase-bigquery_dataset.patch]: This is hard to review. Could we shed a bit more light on what's happening here and where it's coming from?

File: patches/0001-Update-CHANGELOG.md-6336.patch [hunk 2]
```diff
@@ -0,0 +1,88 @@
+From b8fc37e02dde875fb440056ec9489ae1226196fc Mon Sep 17 00:00:00 2001
+From: Thomas Rodgers <thomasrodgers@google.com>
+Date: Mon, 25 Sep 2023 23:29:30 +0000
```
Comment (t0yv0) [inline patches/0001-Update-CHANGELOG.md-6336.patch]: Does this new patch do anything for us?

File: patches/0004-rebase-bigquery_dataset.patch (hunk 1 above)
Comment (guineveresaenger) [reply patches/0004-rebase-bigquery_dataset.patch]: Yeah! basically there was a patch conflict due to upstream adding [these extra lines](https://github.com/hashicorp/terraform-provider-google-beta/blob/main/google-beta/services/bigquery/resource_bigquery_dataset.go#L676) right in the middle of our patch.
Resolving the conflict resulted in this patch.

**Comment by mjeffryes** [comment]
Just a quick comment to note that I've recently rerun the benchmarks for this branch vs master and there do not appear to be any significant performance differences:

```
                                                                                                                    │   old.txt    │                 new.txt                  │
                                                                                                                    │    sec/op    │    sec/op      vs base                   │
/newPlugin/gcp                                                                                                        212.9m ±  0%    208.5m ±  0%    -2.09% (n=220+210)
/newPlugin/yaml:/Users/mjeffryes/workspace/pulumi/gcp/.pulumi/bin/pulumi-language-yaml                                16.21m ±  1%    15.35m ±  1%    -5.27% (n=210+200)

/bigquery-table/ResourceProvider/CheckConfig/pulumi:providers:gcp                                     243.5m ±  4%    263.6m ±  9%    +8.26% (p=0.035 n=10)
/bigquery-datasetaccess-3/ResourceProvider/CheckConfig/pulumi:providers:gcp                           235.7m ±  5%    253.6m ±  8%    +7.61% (p=0.029 n=10)
/compute-disk-3/ResourceProvider/CheckConfig/pulumi:providers:gcp                                     239.3m ±  6%    252.8m ±  9%    +5.63% (p=0.043 n=10)
/storage-bucketobject/ResourceProvider/CheckConfig/pulumi:providers:gcp                               235.0m ±  3%    250.9m ±  9%    +6.74% (p=0.001 n=10)

/bigquery-datasetaccess-3/ResourceMonitor/RegisterResource/pulumi:providers:gcp                       448.9m ±  3%    465.5m ±  4%    +3.69% (p=0.011 n=10)
/cloudfunctions-function/ResourceMonitor/RegisterResource/gcp:storage/bucket:Bucket                   148.8m ±  6%    160.5m ± 10%    +7.89% (p=0.005 n=10)
/compute-disk-1/ResourceMonitor/RegisterResource/gcp:compute/disk:Disk                                153.5m ±  8%    167.3m ±  6%    +9.03% (p=0.019 n=10)
/dns-recordset/ResourceMonitor/RegisterResource/gcp:dns/recordSet:RecordSet                           150.8m ±  5%    162.7m ±  8%    +7.91% (p=0.007 n=10)
/pubsub-topic/ResourceMonitor/RegisterResource/gcp:pubsub/topic:Topic                                 155.8m ±  6%    161.5m ± 32%    +3.66% (p=0.011 n=10)
/storage-bucketobject/ResourceMonitor/RegisterResource/pulumi:providers:gcp                           450.0m ±  2%    457.8m ±  5%    +1.75% (p=0.007 n=10)
```

The smattering of differences in CheckConfig and RegisterResource I think are mostly artifacts of the number of different metrics collected (you test enough things, you find an effect somewhere https://xkcd.com/882/); I've spot checked a few of these by running the benchmarks again and they don't reappear.

The *very* slight decrease in plugin load time was stable.

Anyway, no performance concerns for the release. (yay!)

**Pull request merged**